File: Visual/XSharp/Core/CorePrep/Wire/Encode.cpp
// SPDX-FileCopyrightText: 2026 Leitwolf <support@xsharp-lang.xyz>
// SPDX-License-Identifier: MPL-2.0

#include "Visual/XSharp/Core/CorePrep/Wire.hpp"

#include <limits>
#include <string_view>
#include <type_traits>

namespace visual_xsharp::core::wire
{
namespace
{
class Writer final
{
public:
    explicit Writer(const Limits &limits) : limits_(limits) {}

    [[nodiscard]] auto finish() && -> EncodeResult
    {
        if(!error_ && bytes_.size() > limits_.maximum_wire_bytes)
            fail(ErrorKind::LimitExceeded, "wire byte length", "encoded document exceeds configured byte limit");
        return EncodeResult{std::move(bytes_), std::move(error_)};
    }

    void document(const CorePrepModule &module)
    {
        for(const auto value : magic)
            byte(value);
        unsigned_integer(current_version);
        unsigned_integer<std::uint16_t>(0);
        qualified_name(module.name, "module name");
        vector(module.functions, limits_.maximum_functions, "function count",
               [this](const Function &function) { this->function(function); });
    }

private:
    const Limits &limits_;
    std::vector<std::uint8_t> bytes_;
    std::optional<Error> error_;

    void fail(ErrorKind kind, std::string context, std::string message)
    {
        if(!error_)
            error_ = Error{kind, bytes_.size(), std::move(context), std::move(message)};
    }

    void byte(std::uint8_t value)
    {
        if(!error_)
            bytes_.push_back(value);
    }

    template <typename Integer>
    void unsigned_integer(Integer value)
    {
        static_assert(std::is_unsigned_v<Integer>);
        for(std::size_t shift = 0; shift < sizeof(Integer) * 8U; shift += 8U)
            byte(static_cast<std::uint8_t>((value >> shift) & static_cast<Integer>(0xffU)));
    }

    void count(std::size_t value, std::size_t maximum, std::string_view context)
    {
        if(value > maximum || value > std::numeric_limits<std::uint32_t>::max())
        {
            fail(ErrorKind::LimitExceeded, std::string(context), "collection count exceeds wire limit");
            return;
        }
        unsigned_integer(static_cast<std::uint32_t>(value));
    }

    template <typename Value, typename Encode>
    void vector(const std::vector<Value> &values, std::size_t maximum, std::string_view context, Encode encode)
    {
        count(values.size(), maximum, context);
        if(error_)
            return;
        for(const auto &value : values)
        {
            encode(value);
            if(error_)
                return;
        }
    }

    void text(const std::u32string &value, std::string_view context)
    {
        count(value.size(), limits_.maximum_string_code_points, context);
        for(const auto code_point : value)
        {
            const auto numeric = static_cast<std::uint32_t>(code_point);
            if(numeric > 0x10ffffU || (numeric >= 0xd800U && numeric <= 0xdfffU))
            {
                fail(ErrorKind::InvalidCodePoint, std::string(context), "text contains a non-scalar Unicode code point");
                return;
            }
            unsigned_integer(numeric);
        }
    }

    void qualified_name(const std::vector<std::u32string> &parts, std::string_view context)
    {
        vector(parts, 65535U, context, [this, context](const std::u32string &part) { text(part, context); });
    }

    void symbol(const SymbolName &name, std::string_view context)
    {
        if(name.id == 0)
        {
            fail(ErrorKind::InvalidSymbol, std::string(context), "symbol id must be positive");
            return;
        }
        unsigned_integer(name.id);
        text(name.spelling, context);
    }

    void type(const Type &value, std::size_t depth = 0)
    {
        if(depth > limits_.maximum_type_depth)
        {
            fail(ErrorKind::LimitExceeded, "type", "type nesting exceeds configured limit");
            return;
        }
        byte(static_cast<std::uint8_t>(value.kind));
        switch(value.kind)
        {
        case Type::Kind::Unit:
        case Type::Kind::Bool:
        case Type::Kind::Int64:
        case Type::Kind::Int32:
        case Type::Kind::String: return;
        case Type::Kind::Function:
            if(value.components.empty())
            {
                fail(ErrorKind::UnsupportedType, "function type", "function type has no result component");
                return;
            }
            count(value.components.size() - 1U, 65535U, "function type parameter count");
            for(std::size_t index = 0; index + 1U < value.components.size(); ++index)
                type(value.components[index], depth + 1U);
            type(value.components.back(), depth + 1U);
            return;
        case Type::Kind::Named:
            qualified_name(value.name, "named type");
            vector(value.components, 65535U, "type argument count",
                   [this, depth](const Type &argument) { type(argument, depth + 1U); });
            return;
        case Type::Kind::TypeVariable:
            symbol(value.variable, "type variable symbol");
            return;
        }
    }

    void literal(const Atom &value)
    {
        switch(value.type.kind)
        {
        case Type::Kind::Unit:
            if(!std::holds_alternative<std::monostate>(value.literal))
                fail(ErrorKind::UnsupportedType, "unit literal", "literal payload does not match unit type");
            return;
        case Type::Kind::Bool:
            if(const auto *boolean = std::get_if<bool>(&value.literal))
                byte(*boolean ? 1U : 0U);
            else
                fail(ErrorKind::UnsupportedType, "bool literal", "literal payload does not match bool type");
            return;
        case Type::Kind::Int64:
            if(const auto *integer = std::get_if<std::int64_t>(&value.literal))
                unsigned_integer(static_cast<std::uint64_t>(*integer));
            else
                fail(ErrorKind::UnsupportedType, "int64 literal", "literal payload does not match int64 type");
            return;
        case Type::Kind::Int32:
            if(const auto *integer = std::get_if<std::int32_t>(&value.literal))
                unsigned_integer(static_cast<std::uint32_t>(*integer));
            else
                fail(ErrorKind::UnsupportedType, "int32 literal", "literal payload does not match int32 type");
            return;
        case Type::Kind::String:
            if(const auto *string = std::get_if<std::u32string>(&value.literal))
                text(*string, "string literal");
            else
                fail(ErrorKind::UnsupportedType, "string literal", "literal payload does not match string type");
            return;
        case Type::Kind::Function:
        case Type::Kind::Named:
        case Type::Kind::TypeVariable:
            fail(ErrorKind::UnsupportedType, "literal", "non-primitive literal types cannot cross CorePrep wire");
            return;
        }
    }

    void atom(const Atom &value)
    {
        byte(static_cast<std::uint8_t>(value.kind));
        type(value.type);
        if(value.kind == Atom::Kind::Variable)
            symbol(value.symbol, "variable symbol");
        else
            literal(value);
    }

    void operation(Operation operation, const std::vector<Atom> &operands)
    {
        byte(static_cast<std::uint8_t>(operation));
        vector(operands, limits_.maximum_operands_per_instruction, "operand count",
               [this](const Atom &operand) { atom(operand); });
    }

    void instruction(const Instruction &value)
    {
        byte(static_cast<std::uint8_t>(value.kind));
        switch(value.kind)
        {
        case Instruction::Kind::Bind:
            symbol(value.destination, "binding symbol");
            type(value.type);
            byte(value.mutable_binding ? 1U : 0U);
            operation(value.operation, value.operands);
            return;
        case Instruction::Kind::Assign:
            symbol(value.destination, "assignment symbol");
            if(value.operands.size() != 1U)
            {
                fail(ErrorKind::InvalidCount, "assignment", "assignment must contain exactly one atom");
                return;
            }
            atom(value.operands.front());
            return;
        case Instruction::Kind::Evaluate:
            operation(value.operation, value.operands);
            return;
        }
    }

    void terminator(const Terminator &value)
    {
        byte(static_cast<std::uint8_t>(value.kind));
        switch(value.kind)
        {
        case Terminator::Kind::Return: atom(value.value); return;
        case Terminator::Kind::Branch:
            atom(value.value);
            unsigned_integer(value.true_target);
            unsigned_integer(value.false_target);
            return;
        case Terminator::Kind::Jump: unsigned_integer(value.true_target); return;
        case Terminator::Kind::Unreachable: return;
        }
    }

    void block(const Block &value)
    {
        unsigned_integer(value.id);
        vector(value.instructions, limits_.maximum_instructions_per_block, "instruction count",
               [this](const Instruction &item) { instruction(item); });
        terminator(value.terminator);
    }

    void parameter(const Parameter &value)
    {
        symbol(value.symbol, "parameter symbol");
        type(value.type);
    }

    void function(const Function &value)
    {
        symbol(value.symbol, "function symbol");
        vector(value.parameters, limits_.maximum_parameters_per_function, "parameter count",
               [this](const Parameter &item) { parameter(item); });
        type(value.return_type);
        unsigned_integer(value.entry);
        vector(value.blocks, limits_.maximum_blocks_per_function, "block count",
               [this](const Block &item) { block(item); });
    }
};
} // namespace

auto encode(const CorePrepModule &module, const Limits &limits) -> EncodeResult
{
    Writer writer(limits);
    writer.document(module);
    return std::move(writer).finish();
}
} // namespace visual_xsharp::core::wire

File: Compiler/Haskell/Core/src/Visual/XSharp/Core/CorePrep/Wire/Encode.hs
-- SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
-- SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0
module Visual.XSharp.Core.CorePrep.Wire.Encode (encodeCorePrep, encodeCorePrepWith) where

import Data.Bits (Bits, shiftR, (.&.))
import Data.Char (ord)
import Data.List (unfoldr)
import Data.Word (Word16, Word32, Word64, Word8)
import Visual.XSharp.AST
import Visual.XSharp.Core
import Visual.XSharp.Core.CorePrep
import Visual.XSharp.Core.CorePrep.Wire.Format

type Encoder = Either WireError [Word8]

encodeCorePrep :: CorePrepModule -> Either WireError [Word8]
encodeCorePrep = encodeCorePrepWith defaultWireLimits

encodeCorePrepWith :: WireLimits -> CorePrepModule -> Either WireError [Word8]
encodeCorePrepWith limits moduleValue = do
    payload <- encodeModule limits moduleValue
    let bytes = wireMagic ++ word16 (wireVersionNumber currentWireVersion) ++ word16 0 ++ payload
    requireLimit limits "wire byte length" (maximumWireBytes limits) (length bytes)
    pure bytes

encodeModule :: WireLimits -> CorePrepModule -> Encoder
encodeModule limits moduleValue = do
    name <- encodeQualifiedName limits (corePrepModuleName moduleValue)
    functions <-
        encodeVector
            limits
            "function count"
            (maximumFunctions limits)
            (encodeFunction limits)
            (corePrepModuleFunctions moduleValue)
    pure (name ++ functions)

encodeFunction :: WireLimits -> CorePrepFunction -> Encoder
encodeFunction limits function = do
    name <- encodeResolvedName limits "function symbol" (corePrepFunctionName function)
    parameters <-
        encodeVector
            limits
            "parameter count"
            (maximumParametersPerFunction limits)
            (encodeParameter limits)
            (corePrepFunctionParameters function)
    result <- encodeType limits (corePrepFunctionReturnType function)
    entry <- encodeNonNegative32 "entry block" (corePrepFunctionEntry function)
    blocks <-
        encodeVector
            limits
            "block count"
            (maximumBlocksPerFunction limits)
            (encodeBlock limits)
            (corePrepFunctionBlocks function)
    pure (name ++ parameters ++ result ++ entry ++ blocks)

encodeParameter :: WireLimits -> (ResolvedName, Type) -> Encoder
encodeParameter limits (name, valueType) = do
    encodedName <- encodeResolvedName limits "parameter symbol" name
    encodedType <- encodeType limits valueType
    pure (encodedName ++ encodedType)

encodeBlock :: WireLimits -> CorePrepBlock -> Encoder
encodeBlock limits block = do
    identifier <- encodeNonNegative32 "block id" (corePrepBlockId block)
    instructions <-
        encodeVector
            limits
            "instruction count"
            (maximumInstructionsPerBlock limits)
            (encodeInstruction limits)
            (corePrepBlockInstructions block)
    terminator <- encodeTerminator limits (corePrepBlockTerminator block)
    pure (identifier ++ instructions ++ terminator)

encodeInstruction :: WireLimits -> CorePrepInstruction -> Encoder
encodeInstruction limits instruction = case instruction of
    CorePrepBind name valueType mutable operation -> do
        encodedName <- encodeResolvedName limits "binding symbol" name
        encodedType <- encodeType limits valueType
        encodedOperation <- encodeOperation limits operation
        pure ([0] ++ encodedName ++ encodedType ++ encodeBool mutable ++ encodedOperation)
    CorePrepAssign name atom -> do
        encodedName <- encodeResolvedName limits "assignment symbol" name
        encodedAtom <- encodeAtom limits atom
        pure ([1] ++ encodedName ++ encodedAtom)
    CorePrepEvaluate operation -> do
        encodedOperation <- encodeOperation limits operation
        pure ([2] ++ encodedOperation)

encodeOperation :: WireLimits -> CorePrepOperation -> Encoder
encodeOperation limits operation = case operation of
    CorePrepCopy atom -> encodeTaggedAtoms limits 0 [atom]
    CorePrepCall callee arguments -> encodeTaggedAtoms limits 1 (callee : arguments)
    CorePrepPrimitive primitive arguments -> encodeTaggedAtoms limits (primitiveTag primitive) arguments
    CorePrepMakeClosure function captures -> do
        encodedFunction <- encodeResolvedName limits "closure function symbol" function
        encodedCaptures <-
            encodeVector
                limits
                "closure capture count"
                (maximumOperandsPerInstruction limits)
                (encodeCapture limits)
                captures
        pure ([18] ++ encodedFunction ++ encodedCaptures)

encodeCapture :: WireLimits -> CorePrepCapture -> Encoder
encodeCapture limits (CorePrepCapture mode name valueType atom) = do
    encodedName <- encodeResolvedName limits "capture symbol" name
    encodedType <- encodeType limits valueType
    encodedAtom <- encodeAtom limits atom
    pure (captureModeTag mode : encodedName ++ encodedType ++ encodedAtom)

captureModeTag :: CaptureMode -> Word8
captureModeTag mode = case mode of
    StrongCapture -> 0
    WeakCapture -> 1
    UnownedCapture -> 2

encodeTaggedAtoms :: WireLimits -> Word8 -> [CorePrepAtom] -> Encoder
encodeTaggedAtoms limits tag atoms = do
    encoded <- encodeVector limits "operand count" (maximumOperandsPerInstruction limits) (encodeAtom limits) atoms
    pure (tag : encoded)

encodeAtom :: WireLimits -> CorePrepAtom -> Encoder
encodeAtom limits atom = case atom of
    CorePrepVariable name valueType -> do
        encodedName <- encodeResolvedName limits "variable symbol" name
        encodedType <- encodeType limits valueType
        pure ([0] ++ encodedType ++ encodedName)
    CorePrepLiteral literal valueType -> do
        encodedType <- encodeType limits valueType
        encodedLiteral <- encodeLiteral limits valueType literal
        pure ([1] ++ encodedType ++ encodedLiteral)

encodeLiteral :: WireLimits -> Type -> CoreLiteral -> Encoder
encodeLiteral limits valueType literal = case (primitiveTypeTag valueType, literal) of
    (Just 0, CoreUnit) -> pure []
    (Just 1, CoreBoolean value) -> pure (encodeBool value)
    (Just tag, CoreInteger value)
        | integerWireTag tag || tag == 8 -> encodeInteger limits value
    (Just tag, CoreFloating spelling)
        | floatingWireTag tag -> encodeAscii limits "floating literal" spelling
    (Just 4, CoreString value) -> encodeText limits "string literal" value
    _ -> Left (wireError UnsupportedType 0 "literal" "literal payload does not match its declared type")

encodeTerminator :: WireLimits -> CorePrepTerminator -> Encoder
encodeTerminator limits terminator = case terminator of
    CorePrepReturn atom -> do encoded <- encodeAtom limits atom; pure (0 : encoded)
    CorePrepBranch atom trueTarget falseTarget -> do
        encoded <- encodeAtom limits atom
        trueBlock <- encodeNonNegative32 "true target" trueTarget
        falseBlock <- encodeNonNegative32 "false target" falseTarget
        pure ([1] ++ encoded ++ trueBlock ++ falseBlock)
    CorePrepJump target -> do encoded <- encodeNonNegative32 "jump target" target; pure (2 : encoded)
    CorePrepUnreachable -> pure [3]

encodeResolvedName :: WireLimits -> String -> ResolvedName -> Encoder
encodeResolvedName limits context name = do
    symbol <- encodePositive64 context (symbolIdValue (resolvedSymbol name))
    spelling <- encodeText limits (context ++ " spelling") (identifierText (resolvedSpelling name))
    pure (symbol ++ spelling)

encodeQualifiedName :: WireLimits -> QualifiedName -> Encoder
encodeQualifiedName limits (QualifiedName parts) =
    encodeVector
        limits
        "qualified name part count"
        65535
        (encodeText limits "qualified name part" . identifierText)
        parts

encodeType :: WireLimits -> Type -> Encoder
encodeType limits = encodeTypeAt limits 0

encodeTypeAt :: WireLimits -> Int -> Type -> Encoder
encodeTypeAt limits depth valueType
    | depth > maximumTypeDepth limits = Left (wireError LimitExceeded 0 "type" "type nesting exceeds configured limit")
    | Just tag <- primitiveTypeTag valueType = pure [tag]
    | NamedType name arguments <- valueType = do
        encodedName <- encodeQualifiedName limits name
        encodedArguments <- encodeVector limits "type argument count" 65535 (encodeTypeAt limits (depth + 1)) arguments
        pure ([6] ++ encodedName ++ encodedArguments)
    | FunctionType parameters result <- valueType = do
        encodedParameters <-
            encodeVector
                limits
                "function type parameter count"
                65535
                (encodeTypeAt limits (depth + 1))
                parameters
        encodedResult <- encodeTypeAt limits (depth + 1) result
        pure ([5] ++ encodedParameters ++ encodedResult)
    | TypeVariable name <- valueType = do
        encodedName <- encodeResolvedName limits "type variable symbol" name
        pure ([7] ++ encodedName)
    | ErrorType <- valueType =
        Left (wireError UnsupportedType 0 "type" "unresolved ErrorType cannot cross the CorePrep boundary")

primitiveTypeTag :: Type -> Maybe Word8
primitiveTypeTag valueType
    | valueType == unitType = Just 0
    | valueType == boolType = Just 1
    | valueType == stringType = Just 4
    | Just tag <- scalarTypeTag valueType = Just tag
    | otherwise = Nothing

-- Tags 0-7 preserve the v2 structural layout. New scalar tags are append-only,
-- which keeps corrupted-version diagnostics deterministic and makes wire dumps
-- easier to compare while the compiler remains pre-1.0.
scalarTypeTag :: Type -> Maybe Word8
scalarTypeTag (NamedType (QualifiedName [Identifier name]) []) = lookup name scalarTags
scalarTypeTag _ = Nothing

scalarTags :: [(String, Word8)]
scalarTags =
    [ ("bool", 1)
    , ("int", 2)
    , ("long", 3)
    , ("char", 8)
    , ("byte", 9)
    , ("short", 10)
    , ("longint", 11)
    , ("ubyte", 12)
    , ("ushort", 13)
    , ("ulong", 14)
    , ("uint", 15)
    , ("ulongint", 16)
    , ("sfloat", 17)
    , ("lfloat", 18)
    , ("float", 19)
    , ("double", 20)
    ]

integerWireTag :: Word8 -> Bool
integerWireTag tag = tag `elem` [2, 3, 9, 10, 11, 12, 13, 14, 15, 16]

floatingWireTag :: Word8 -> Bool
floatingWireTag tag = tag `elem` [17, 18, 19, 20]

primitiveTag :: CorePrimitive -> Word8
primitiveTag primitive = case primitive of
    CoreAdd -> 2
    CoreSubtract -> 3
    CoreMultiply -> 4
    CoreDivide -> 5
    CoreFloorDivide -> 6
    CoreRemainder -> 7
    CoreLessThan -> 8
    CoreLessEqual -> 9
    CoreGreaterThan -> 10
    CoreGreaterEqual -> 11
    CoreEqual -> 12
    CoreNotEqual -> 13
    CoreLogicalAnd -> 14
    CoreLogicalOr -> 15
    CoreNegate -> 16
    CoreLogicalNot -> 17

encodeVector :: WireLimits -> String -> Int -> (a -> Encoder) -> [a] -> Encoder
encodeVector limits context maximumCount encode values = do
    requireLimit limits context maximumCount (length values)
    encodedCount <- encodeNonNegative32 context (length values)
    encodedValues <- traverse encode values
    pure (encodedCount ++ concat encodedValues)

encodeText :: WireLimits -> String -> String -> Encoder
encodeText limits context value = do
    requireLimit limits context (maximumStringCodePoints limits) (length value)
    count <- encodeNonNegative32 context (length value)
    codePoints <- traverse encodeCodePoint value
    pure (count ++ concat codePoints)
    where
        encodeCodePoint character
            | code >= 0xD800 && code <= 0xDFFF =
                Left (wireError InvalidCodePoint 0 context "surrogate code point is not a scalar value")
            | code > 0x10FFFF = Left (wireError InvalidCodePoint 0 context "code point exceeds Unicode range")
            | otherwise = Right (word32 (fromIntegral code))
            where
                code = ord character

encodeInteger :: WireLimits -> Integer -> Encoder
encodeInteger limits value = do
    let magnitude = integerMagnitude (abs value)
    requireLimit limits "integer magnitude" (maximumNumericBytes limits) (length magnitude)
    count <- encodeNonNegative32 "integer magnitude length" (length magnitude)
    pure (encodeBool (value < 0) ++ count ++ magnitude)

integerMagnitude :: Integer -> [Word8]
integerMagnitude 0 = []
integerMagnitude value = reverse (unfoldr step value)
    where
        step 0 = Nothing
        step remaining = Just (fromIntegral (remaining `mod` 256), remaining `div` 256)

encodeAscii :: WireLimits -> String -> String -> Encoder
encodeAscii limits context value = do
    requireLimit limits context (maximumNumericBytes limits) (length value)
    count <- encodeNonNegative32 (context ++ " length") (length value)
    bytes <- traverse ascii value
    pure (count ++ bytes)
    where
        ascii character
            | ord character <= 0x7f = Right (fromIntegral (ord character))
            | otherwise = Left (wireError InvalidInteger 0 context "numeric spelling must contain ASCII characters only")

encodePositive64 :: String -> Int -> Encoder
encodePositive64 context value
    | value <= 0 = Left (wireError InvalidSymbol 0 context "symbol id must be positive")
    | otherwise = pure (word64 (fromIntegral value))

encodeNonNegative32 :: String -> Int -> Encoder
encodeNonNegative32 context value
    | value < 0 || toInteger value > toInteger (maxBound :: Word32) =
        Left (wireError InvalidCount 0 context "value does not fit unsigned 32-bit wire field")
    | otherwise = pure (word32 (fromIntegral value))

encodeBool :: Bool -> [Word8]
encodeBool False = [0]
encodeBool True = [1]

requireLimit :: WireLimits -> String -> Int -> Int -> Either WireError ()
requireLimit _ context maximumValue actual
    | actual > maximumValue =
        Left (wireError LimitExceeded 0 context ("count " ++ show actual ++ " exceeds limit " ++ show maximumValue))
    | otherwise = Right ()

word16 :: Word16 -> [Word8]
word16 value = [byte value 0, byte value 8]

word32 :: Word32 -> [Word8]
word32 value = [byte value 0, byte value 8, byte value 16, byte value 24]

word64 :: Word64 -> [Word8]
word64 value =
    [ byte value 0
    , byte value 8
    , byte value 16
    , byte value 24
    , byte value 32
    , byte value 40
    , byte value 48
    , byte value 56
    ]

byte :: (Integral a, Bits a) => a -> Int -> Word8
byte value amount = fromIntegral ((value `shiftR` amount) .&. 0xff)
